In [5]:
import numpy as np
import pandas as pd

In [6]:
data={'Mail':['Mail 1','Mail 2','Mail 3','Mail 4'],
    'Dear':['1','1','0','1'],
    'Dinner':['2','0','3','0'],
    'Money':['1','4','0','2']}

In [7]:
df=pd.DataFrame(data)

In [8]:
df

Mail Dear Dinner Money
0  Mail 1    1      2     1
1  Mail 2    1      0     4
2  Mail 3    0      3     0
3  Mail 4    1      0     2

Data is about frequency(cell) of words(columns) present in a mail(row).

# PROBLEM STATEMENT: 

- Need to classify the mail as spam or not-- If the output value for an email is greater than 0.5, then we would classify it as spam, and if it's less than 0.5, we would classify it as not spam.

- For the given data write the forward pass through ANN using python and numpy.

- Pass the data through the network, display the outputs and error.

- Use proper activation functions.
          
- Assume weights and biases.


### Given architecture of the neural network

In [42]:
input_size = 3
output_size = 1
hidden_size = 4

### Initializing weights and biases

In [87]:
weights_input_hidden = np.random.normal(0.0, input_size ** -0.5, (input_size, hidden_size))
bias_hidden = np.zeros((1, hidden_size))
weights_hidden_output = np.random.normal(0.0, hidden_size ** -0.5, (hidden_size, output_size))
bias_output = np.zeros((1, output_size))

In [ ]:
inputs = np.array([[1, 2, 1],
                   [1, 0, 4],
                   [0, 3, 0],
                   [1, 0, 2]])

In [ ]:
expected_output = np.array([[0],
                            [1],
                            [0],
                            [1]])

Where 0 - Not Spam and 1 - Spam

### Sigmoid activation function is used

In [77]:
def sigmoid(inputs):
    output = 1/(1+np.exp(-np.array(inputs)))
    return output

In [95]:
print(sigmoid(output))

[[0.97587298]
 [0.99183743]
 [0.93086158]
 [0.97811873]]


### INFERENCE

We can conclude that Mail2 has the highest probability of being spam (0.9918), followed by Mail4 (0.9781), Mail1 (0.9759), and Mail3 (0.9309). 

### Train the neural network:-

-Performing forward pass through the network

-To get the outputs

In [91]:
hidden_layer = sigmoid(np.dot(inputs, weights_input_hidden) + bias_hidden)
output_layer = sigmoid(np.dot(hidden_layer, weights_hidden_output) + bias_output)

In [92]:
print(output_layer)

[[0.32673222]
 [0.41110417]
 [0.29210546]
 [0.38764546]]


### INFERENCE

We can conclude that the output values represent the probability of each data point belonging to the positive class (spam in this case), **Mail2** has the highest probability of being spam (0.4111), followed by Mail4 (0.3876), Mail1 (0.3267), and Mail3 (0.2921). 

### Calculating the error   

In [93]:
output_error = expected_output - output_layer
hidden_error = np.dot(output_error, weights_hidden_output.T) * hidden_layer * (1 - hidden_layer)

In [94]:
print(output_error)

[[-0.32673222]
 [ 0.58889583]
 [-0.29210546]
 [ 0.61235454]]


### INFERENCE

From the above values we conclude that the values are not probabilities, as they are not bounded between 0 and 1. Instead, they are the raw output of the neural network's final layer, which means that they can be positive or negative.

Therefore, **Mail2** is **0.58889583** and **Mail4** is **0.61235454** which is also greater than 0.5 that are classified as **spam** by the neural network and **Mail1** is **-0.32673222** and **Mail3** is **-0.29210546** have negative output values that is **not spam**.